File: src/selection.js
'use strict';

function tournamentSelection(population, rng, tournamentSize = 3) {
  let best = null;
  for (let i = 0; i < tournamentSize; i++) {
    const idx = rng.nextInt(0, population.length);
    const candidate = population[idx];
    if (best === null || candidate.fitness > best.fitness) {
      best = candidate;
    }
  }
  return best;
}

function rouletteSelection(population, rng) {
  const minFitness = Math.min(...population.map((o) => o.fitness));
  const offset = minFitness < 0 ? -minFitness : 0;
  const totalFitness = population.reduce((s, o) => s + o.fitness + offset, 0);
  if (totalFitness === 0) {
    return population[rng.nextInt(0, population.length)];
  }
  let spin = rng.next() * totalFitness;
  for (const org of population) {
    spin -= org.fitness + offset;
    if (spin <= 0) return org;
  }
  return population[population.length - 1];
}

function rankSelection(population, rng) {
  const sorted = [...population].sort((a, b) => a.fitness - b.fitness);
  const n = sorted.length;
  const totalRank = (n * (n + 1)) / 2;
  let spin = rng.next() * totalRank;
  for (let i = 0; i < n; i++) {
    spin -= i + 1;
    if (spin <= 0) return sorted[i];
  }
  return sorted[n - 1];
}

function eliteSelection(population, count) {
  const sorted = [...population].sort((a, b) => b.fitness - a.fitness);
  return sorted.slice(0, count);
}

const STRATEGIES = {
  tournament: tournamentSelection,
  roulette: rouletteSelection,
  rank: rankSelection,
};

module.exports = {
  tournamentSelection,
  rouletteSelection,
  rankSelection,
  eliteSelection,
  STRATEGIES,
};

File: python/phyloid_engine/selection.py
from __future__ import annotations

from .organism import Organism
from .prng import Mulberry32


def tournament_selection(
    population: list[Organism], rng: Mulberry32, tournament_size: int = 3
) -> Organism:
    """Select the fittest organism from a random subset of the population."""
    if not population:
        raise ValueError("population must not be empty")
    if tournament_size < 1:
        raise ValueError(f"tournament_size must be >= 1, got {tournament_size}")
    best: Organism | None = None
    for _ in range(tournament_size):
        idx = rng.next_int(0, len(population))
        candidate = population[idx]
        if best is None or candidate.fitness > best.fitness:
            best = candidate
    assert best is not None
    return best


def roulette_selection(population: list[Organism], rng: Mulberry32) -> Organism:
    """Select an organism with probability proportional to fitness."""
    if not population:
        raise ValueError("population must not be empty")
    min_fitness = min(o.fitness for o in population)
    offset = -min_fitness if min_fitness < 0 else 0
    total_fitness = sum(o.fitness + offset for o in population)
    if total_fitness == 0:
        return population[rng.next_int(0, len(population))]
    spin = rng.next() * total_fitness
    for org in population:
        spin -= org.fitness + offset
        if spin <= 0:
            return org
    return population[-1]


def rank_selection(population: list[Organism], rng: Mulberry32) -> Organism:
    """Select an organism with probability proportional to fitness rank."""
    if not population:
        raise ValueError("population must not be empty")
    sorted_pop = sorted(population, key=lambda o: o.fitness)
    n = len(sorted_pop)
    total_rank = n * (n + 1) // 2
    spin = rng.next() * total_rank
    for i, org in enumerate(sorted_pop):
        spin -= i + 1
        if spin <= 0:
            return org
    return sorted_pop[-1]


def elite_selection(population: list[Organism], count: int) -> list[Organism]:
    """Return the top N organisms sorted by fitness descending."""
    sorted_pop = sorted(population, key=lambda o: o.fitness, reverse=True)
    return sorted_pop[:count]
